File: source/loader/platform/efi/disk.c
/**
* The MIT License (MIT)
*
* Copyright (c) 2014 Gil Mendes
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the "Software"), to deal
* in the Software without restriction, including without limitation the rights
* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
* copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in all
* copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
* SOFTWARE.
*/

/**
 * @file
 * @brief               EFI disk functions.
 */

#include <lib/string.h>

#include <efi/efi.h>

#include <disk.h>
#include <loader.h>
#include <memory.h>

/** Structure containing EFI disk information. */
typedef struct efi_disk {
    disk_device_t disk;                 /**< Disk device header. */

    efi_device_path_t *path;            /**< Device path. */
    efi_block_io_protocol_t *block;     /**< Block I/O protocol. */
    efi_uint32_t media_id;              /**< Media ID. */
} efi_disk_t;

/** Block I/O protocol GUID. */
static efi_guid_t block_io_guid = EFI_BLOCK_IO_PROTOCOL_GUID;

/**
 * Read blocks from an EFI disk.
 *
 * @param  _disk Disk device being read from.
 * @param  buf   Buffer to read into.
 * @param  count Number of blocks to read.
 * @param  lba   BLock number to start reading from.
 * @return       Status code describing the result of operation.
 */
static status_t efi_disk_read_blocks(disk_device_t *_disk, void *buf, size_t count, uint64_t lba) {
    efi_disk_t *disk = (efi_disk_t *)_disk;
    efi_status_t ret;

    ret = efi_call(disk->block->read_blocks, disk->block, disk->media_id, lba, count * disk->disk.block_size, buf);
    return efi_convert_status(ret);
}

/** EFI disk operations structure. */
static disk_ops_t efi_disk_ops = {
    .read_blocks = efi_disk_read_blocks,
};

/** Detect and register all disk devices. */
void efi_disk_init(void) {
    efi_handle_t *handles;
    efi_uintn_t num_handles, num_raw_devices = 0;
    list_t raw_devices, child_devices;
    efi_status_t ret;

    list_init(&raw_devices);
    list_init(&child_devices);

    /* Get a list of all handles supporting the block I/O protocol. */
    ret = efi_locate_handle(EFI_BY_PROTOCOL, &block_io_guid, NULL, &handles, &num_handles);
    if (ret != EFI_SUCCESS) {
        dprintf("efi: no block devices available\n");
        return;
    }

    /*
     * EFI gives us both the raw devices, and any partitions as child devices.
     * We are only interested in the raw devices, as we handle partition maps
     * internally. We want to pick out the raw devices, and identify the type of
     * these devices.
     *
     * It seems like there should be a better way to identify the type, but raw
     * devices don't appear to get flagged with the type of device they are:
     * their device path nodes are just typed as ATA/SCSI/whatever (except for
     * floppies, which can be identified by their ACPI HID). Child devices do
     * get flagged with a device type.
     *
     * What we do then is make a first pass over all devices to get their block
     * protocol. If a device is a raw device (media->logical_partition == 0,
     * then we do some guesswork:
     *
     *  1. If device path node is ACPI, check HID, mark as floppy if matches.
     *  3. Otherwise, if removable, read only, and block size is 2048, mark as CD.
     *  4. Otherwise, mark as HD.
     *
     * We then do a pass over the child devices, and if they identify the type
     * of their parent, then that overrides the type guessed for the raw device.
     */
    for (efi_uintn_t i = 0; i < num_handles; i++) {
        efi_disk_t *disk;
        efi_block_io_media_t *media;

        disk = malloc(sizeof(*disk));
        memset(disk, 0, sizeof(*disk));
        list_init(&disk->disk.device.header);

        disk->path = efi_get_device_path(handles[i]);
        if (!disk->path) {
            free(disk);
            continue;
        }

        ret = efi_open_protocol(handles[i], &block_io_guid, EFI_OPEN_PROTOCOL_GET_PROTOCOL, (void **)&disk->block);
        if (ret != EFI_SUCCESS) {
            dprintf("efi: warning: failed to open block I/O for %pE\n", disk->path);
            free(disk);
            continue;
        }

        media = disk->block->media;

        disk->media_id = media->media_id;
        disk->disk.ops = &efi_disk_ops;
        disk->disk.block_size = media->block_size;
        disk->disk.blocks = media->last_block + 1;

        if (media->logical_partition) {
            list_append(&child_devices, &disk->disk.device.header);
        } else {
            efi_device_path_t *last = efi_last_device_node(disk->path);

            disk->disk.type = DISK_TYPE_HD;
            if (last->type == EFI_DEVICE_PATH_TYPE_ACPI) {
                efi_device_path_acpi_t *acpi = (efi_device_path_acpi_t *)last;

                /* Check EISA ID for a floppy. */
                if (acpi->hid == 0x060441d0)
                    disk->disk.type = DISK_TYPE_FLOPPY;
            } else if (media->removable_media && media->read_only && media->block_size == 2048) {
                disk->disk.type = DISK_TYPE_CDROM;
            }

            list_append(&raw_devices, &disk->disk.device.header);
            num_raw_devices++;
        }
    }

    free(handles);

    /* Pass over child devices to identify their types. */
    list_foreach_safe(&child_devices, iter) {
        efi_disk_t *child = list_entry(iter, efi_disk_t, disk.device.header);
        efi_device_path_t *last = efi_last_device_node(child->path);

        if (last->type == EFI_DEVICE_PATH_TYPE_MEDIA) {
            /* Identify the parent device. */
            list_foreach_safe(&raw_devices, piter) {
                efi_disk_t *parent = list_entry(piter, efi_disk_t, disk.device.header);

                if (efi_is_child_device_node(parent->path, child->path)) {
                    switch (last->subtype) {
                    case EFI_DEVICE_PATH_MEDIA_SUBTYPE_HD:
                        parent->disk.type = DISK_TYPE_HD;
                        break;
                    case EFI_DEVICE_PATH_MEDIA_SUBTYPE_CDROM:
                        parent->disk.type = DISK_TYPE_CDROM;
                        break;
                    }
                }
            }
        }

        list_remove(&child->disk.device.header);
        free(child);
    }

    dprintf("efi: found %zu raw block devices (%zu total):\n", num_raw_devices, num_handles);

    /* Finally add the raw devices. */
    list_foreach_safe(&raw_devices, iter) {
        efi_disk_t *disk = list_entry(iter, efi_disk_t, disk.device.header);

        list_remove(&disk->disk.device.header);
        disk_device_register(&disk->disk);
        dprintf(" %-7s -> %pE\n", disk->disk.device.name, disk->path);
    }
}

File: src/platform/efi/disk.rs
use alloc::vec::Vec;
use core::{cell::UnsafeCell, ptr};

use uefi::{
    proto::device_path::{AcpiDevicePath, DevicePath, DeviceType},
    proto::media::block::BlockIO,
    Handle,
};

use crate::disk::{Disk, DiskType};

use super::{
    device::{is_child_device_node, last_device_node},
    get_loaded_image, get_system_table,
};

/// Size of a CD sector
const CD_SECTOR_SIZE: u32 = 2048;

/// floppy HID
const FLOPPY_HUI: u32 = 0x060441d0;

/// Structure containing EFI disk information
struct EfiDisk<'a> {
    /// Handle to disk
    handle: Handle,
    /// Device path
    path: &'a UnsafeCell<DevicePath>,
    /// Block I/O protocol
    block: &'a UnsafeCell<BlockIO>,
    /// Media ID
    media_id: u32,
    /// Whether the device is the boot device
    boot: bool,
    /// Lba of the boot partition
    boot_partition_lba: u64,
    /// Common disk structure
    disk: Option<Disk>,
}

impl<'a> EfiDisk<'a> {
    /// Get a reference to the efi disk's handle.
    fn handle(&self) -> &Handle {
        &self.handle
    }

    /// Get a reference to the efi disk's path.
    fn path(&self) -> &'a UnsafeCell<DevicePath> {
        &self.path
    }

    /// Get a reference to the efi disk's block.
    fn block(&self) -> &'a UnsafeCell<BlockIO> {
        &self.block
    }

    /// Get a reference to the efi disk's media id.
    fn media_id(&self) -> u32 {
        self.media_id
    }

    /// Get a reference to the efi disk's boot.
    fn is_boot(&self) -> bool {
        self.boot
    }

    /// Get a reference to the efi disk's boot partition lba.
    fn boot_partition_lba(&self) -> u64 {
        self.boot_partition_lba
    }

    /// Get common disk structure
    fn get_common_disk(&self) -> Option<&Disk> {
        self.disk.as_ref()
    }

    /// Set the common disk structure
    fn set_common_disk(&mut self, disk: Option<Disk>) {
        self.disk = disk
    }
}

pub struct EfiDiskManager {}

/// Detect and register all disk devices.
pub fn init() {
    use uefi::ResultExt;

    let mut raw_devices: Vec<EfiDisk> = Vec::new();
    let mut child_devices: Vec<EfiDisk> = Vec::new();

    let bt = get_system_table().boot_services();

    // Get a list of all handles supporting the block I/O protocol.
    let handles = bt
        .find_handles::<BlockIO>()
        .expect_success("efi: no block devices available");

    info!("efi: number of block devices: {}", handles.len());

    // EFI gives us both raw devices, and any partitions as child devices. We are only interested in the raw devices, as
    // we handle partition maps internally. We want to pick out the raw devices, and identify the type of these devices.
    //
    // It seems like there should be a better way to identify the type, but raw devices don't appear to get flagged with
    // the type of the device they are: their path nodes are just typed as ATA/SCSI/whatever (except for floppies, which
    // can be identified by their ACPI HID). Child devices do get flagged with a device type.
    //
    // What we do then is make a first pass over all devices to get their block protocol. If a device is a raw device (
    // is_logical_partition() == true), then we do some guesswork:
    //
    // 1. If device path node is ACPI, check HID, mark as floppy if matches.
    // 2. Otherwise, if removable, read-only, and block size if 2048, mark as CD.
    // 3. Otherwise, mark as HD.
    //
    // We then do a pass ovr the child devices, and if they identify the type of the their parent, then that overrides
    // the type guessed for the raw device.
    handles.iter().for_each(|&handle| {
        let block_cell = bt
            .handle_protocol::<BlockIO>(handle)
            .expect("efi: warning: failed to open block I/O")
            .unwrap();

        unsafe {
            let block = &*block_cell.get();
            let media = block.media();

            // Get device path and ignore the end of hardware path
            let device_path_cell = bt
                .handle_protocol::<DevicePath>(handle)
                .expect("efi: failed to retrieve `DevicePath` protocol from block image handler")
                .unwrap();
            let device_path = &mut *device_path_cell.get();
            match device_path.device_type {
                DeviceType::End => return,
                _ => {}
            }

            // TODO: fix the boot device detection logic
            let loaded_image_device = get_loaded_image().device();

            info!("1>> {:x}", (&loaded_image_device as *const _ as usize));
            let p = 50;
            info!("2>> {:x}", (&p as *const _ as usize));

            // create the new disk
            let mut disk = EfiDisk {
                handle,
                path: device_path_cell,
                block: block_cell,
                media_id: media.media_id(),
                boot: &loaded_image_device as *const _ == &handle as *const _,
                boot_partition_lba: if media.is_media_preset() {
                    media.last_block() + 1
                } else {
                    0
                },
                disk: None,
            };

            if disk.is_boot() {
                info!("efi: boot device is: {:p}", disk.path());
            }

            if media.is_logical_partition() {
                child_devices.push(disk);
            } else {
                let disk_path = &(*disk.path().get());
                let last_node = last_device_node(disk_path);

                let mut common_disk = Disk::new();
                if last_node.device_type == DeviceType::Acpi {
                    let acpi_device =
                        ptr::read(last_node as *const DevicePath as *const AcpiDevicePath);

                    // Check EISA ID for a floppy
                    match acpi_device.hid {
                        FLOPPY_HUI => {
                            common_disk.disk_type = DiskType::Floppy;
                        }
                        _ => {}
                    }
                } else if media.is_removable_media()
                    && media.is_read_only()
                    && media.block_size() == CD_SECTOR_SIZE
                {
                    common_disk.disk_type = DiskType::CDROM;
                }

                disk.set_common_disk(Some(common_disk));
                raw_devices.push(disk);
            }
        }
    });

    // pass over child devices to identify their types
    child_devices.iter().for_each(|child| {
        let child_path = unsafe { &(*child.path().get()) };
        let last_path = last_device_node(child_path);

        // identify the parent device
        raw_devices.iter_mut().for_each(|parent| {
            let parent_path = unsafe { &*parent.path().get() };
            let child_path = unsafe { &*child.path().get() };

            // ignore when there is no relation between the child and the possible parent
            if !is_child_device_node(parent_path, child_path) {
                return;
            }

            // Mark the parent as the boot device if the partition is the boot partition
            if child.is_boot() {
                parent.boot = true;
            }

            if last_path.device_type != DeviceType::Media {
                return;
            }

            unsafe {
                let sub_type = *(&last_path.sub_type as *const _ as *const u8);
                info!(">>> {:x}", sub_type);
            }

            // match last_path.sub_type {
            //     uefi::proto::device_path::DeviceSubType::EndInstance => {}
            //     uefi::proto::device_path::DeviceSubType::EndEntire => {}
            // }
        });
    });
}
